datafusion/physical-expr/src/window/window_frame_state.rs

```diff
@@ -0,0 +1,972 @@
+            .collect::<Result<Vec<ScalarValue>>>()?;
+        let mut start_idx: usize = 0;
+        for _ in 0..self.window_frame_start_idx {
+            let next_group_and_start_index =
+                WindowFrameStateGroups::find_next_group_and_start_index(
+                    range_columns,
+                    &group_values,
+                    start_idx,
+                )?;
+            if let Some(entry) = next_group_and_start_index {
+                (group_values, start_idx) = entry;
+            } else {
+                // not enough groups to generate a window frame
+                self.window_frame_end_idx = self.window_frame_start_idx;
+                self.reached_end = true;
+                return Ok(());
+            }
+        }
+        self.group_start_indices
+            .push_back((group_values, start_idx));
+        self.window_frame_end_idx = self.window_frame_start_idx + 1;
+        Ok(())
+    }
+
+    fn initialized(&self) -> bool {
+        self.reached_end || !self.group_start_indices.is_empty()
+    }
+
+    /// This function advances the window frame by one group.
+    fn advance_one_group<const SEARCH_SIDE: bool>(
+        &mut self,
+        range_columns: &[ArrayRef],
+    ) -> Result<()> {
+        let last_group_values = self.group_start_indices.back();
+        let last_group_values = if let Some(values) = last_group_values {
+            values
+        } else {
+            return Ok(());
+        };
+        let next_group_and_start_index =
+            WindowFrameStateGroups::find_next_group_and_start_index(
+                range_columns,
+                &last_group_values.0,
+                last_group_values.1,
+            )?;
+        if let Some(entry) = next_group_and_start_index {
+            self.group_start_indices.push_back(entry);
+            self.window_frame_end_idx += 1;
+        } else {
+            // not enough groups to proceed
+            self.reached_end = true;
+        }
+        Ok(())
+    }
+
+    /// This function drops the oldest group from the window frame.
+    fn shift_one_group<const SEARCH_SIDE: bool>(&mut self, delta: u64) {
+        let current_window_frame_start_idx = if !SEARCH_SIDE {
+            self.current_group_idx + delta
+        } else if self.current_group_idx >= delta {
+            self.current_group_idx - delta
+        } else {
+            0
+        };
+        if current_window_frame_start_idx > self.window_frame_start_idx {
+            self.group_start_indices.pop_front();
+            self.window_frame_start_idx += 1;
+        }
+    }
+
+    /// This function finds the next group and its start index for a given group and start index.
+    /// It utilizes an exponentially growing step size to find the group boundary.
+    // TODO: For small group sizes, proceeding one-by-one to find the group change can be more efficient.
+    // Statistics about previous group sizes can be used to choose one-by-one vs. exponentially growing,
```

Comment:
I agree there are likely many ways to improve the performance of this algorithm -- however, I think the first thing to do is getting a working implementation that is well tested and structured (as this PR is) and as you say we can drive additional optimizations from there.